src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,197 @@
+            /// <remarks>
+            /// <para>The returned object is an independent reference to the same underlying object. Disposing of the
+            /// returned value multiple times will only cause the reference count to be decreased once.</para>
+            /// </remarks>
+            /// <returns>A new <see cref="ReferenceCountedDisposable{T}"/> pointing to the same underlying object, if it
+            /// has not yet been disposed; otherwise, <see langword="null"/> if this reference to the underlying object
+            /// has already been disposed.</returns>
+            public ReferenceCountedDisposable<T> TryAddReference()
+            {
+                // This value can be set in Dispose. However, even if we have a race condition with Dispose where it's
+                // set to null in the middle of the call, the reference count will be incremented by this method if and
+                // only if a valid new reference was created. This is because the validity of this instance for creating
+                // new references is determined by _referenceCount.
+                var target = _instance;
+                Interlocked.MemoryBarrier();
+                var referenceCount = Volatile.Read(ref _referenceCount);
+
+                // Cannot use Interlocked.Increment because we need to latch the reference count when it reaches 0.
+                while (true)
+                {
+                    var currentValue = Volatile.Read(ref referenceCount.Value);
+                    if (currentValue == 0)
+                    {
+                        // The target is already disposed, and cannot be reused
+                        return null;
+                    }
+
+                    if (Interlocked.CompareExchange(ref referenceCount.Value, currentValue + 1, currentValue) == currentValue)
+                    {
+                        return new ReferenceCountedDisposable<T>(target, referenceCount);
+                    }
+                }
+            }
+
+            public void Dispose()
+            {
+                var referenceCount = Interlocked.Exchange(ref _referenceCount, null);
+                if (referenceCount == null)
+                {
+                    // Already disposed; allow multiple without error.
+                    return;
+                }
+
+                // This is the thread which will write to _instance, so this will act as an atomic read
+                var instance = _instance;
+
+                // Set the instance back to its default value, so in case someone forgets to dispose of one of the
+                // counted references the finalizer of the underlying disposable object will have a chance to clean up
+                // the unmanaged resources as soon as possible.
+                Interlocked.MemoryBarrier();
+                _instance = default(T);
+
+#if !DEBUG
+                var decrementedValue = Interlocked.Decrement(ref referenceCount.Value);
```

Comment:
Having different codepaths for release and debug in such a delicate codepath is concerning.  It is not at all clear to me that these codepaths are doing the same thing, and this may make things very hard to track down if we encounter issues that only occur in one flavor and not hte other.